src/query/functions/temporal/aggregation.go

```diff
@@ -0,0 +1,161 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+)
+
+const (
+	// AvgTemporalType calculates the average of all values in the specified interval
+	AvgTemporalType = "avg_over_time"
+
+	// CountTemporalType calculates count of all values in the specified interval
+	CountTemporalType = "count_over_time"
+
+	// MinTemporalType calculates the minimum of all values in the specified interval
+	MinTemporalType = "min_over_time"
+
+	// MaxTemporalType calculates the maximum of all values in the specified interval
+	MaxTemporalType = "max_over_time"
+
+	// SumTemporalType calculates the sum of all values in the specified interval
+	SumTemporalType = "sum_over_time"
+
+	// StdDevTemporalType calculates the standard deviation of all values in the specified interval
+	StdDevTemporalType = "stddev_over_time"
+
+	// StdVarTemporalType calculates the standard variance of all values in the specified interval
+	StdVarTemporalType = "stdvar_over_time"
+)
+
+var (
+	aggFuncs = map[string]interface{}{
+		AvgTemporalType:    struct{}{},
+		CountTemporalType:  struct{}{},
+		MinTemporalType:    struct{}{},
+		MaxTemporalType:    struct{}{},
+		SumTemporalType:    struct{}{},
+		StdDevTemporalType: struct{}{},
+		StdVarTemporalType: struct{}{},
+	}
+)
+
+// NewAggOp creates a new base temporal transform with a specified node
+func NewAggOp(args []interface{}, optype string) (transform.Params, error) {
+	if _, ok := aggFuncs[optype]; !ok {
+		return emptyOp, fmt.Errorf("unknown aggregation type: %s", optype)
+	}
+
+	return newBaseOp(args, optype, newAggNode)
+}
+
+func newAggNode(op baseOp, controller *transform.Controller) Processor {
+	return &aggNode{
+		op:         op,
+		controller: controller,
+	}
+}
+
+type aggNode struct {
+	op         baseOp
+	controller *transform.Controller
+}
+
+func (a *aggNode) Process(values []float64) float64 {
+	switch a.op.operatorType {
+	case AvgTemporalType:
+		var mean, count float64
+		for _, v := range values {
+			if !math.IsNaN(v) {
+				count++
+				mean += (v - mean) / count
```

Comment:
Why do you need a rolling average here? Can't you just keep track of sum and count then return sum/count? You'll also need to account for cases where count == 0

Reply:
Was following how Prom did it, but yeah we can just do that instead.